File: rag-server/internal/auth/token_service.go
package auth

import (
	"fmt"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

// TokenPair represents a pair of Public and Access tokens
type TokenPair struct {
	PublicToken  string `json:"public_token"`
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	TokenType    string `json:"token_type"`
	ExpiresIn    int64  `json:"expires_in"`
}

// Claims represents JWT access token claims
type Claims struct {
	UserID   string   `json:"user_id"`
	Email    string   `json:"email"`
	Roles    []string `json:"roles"`
	Service  string   `json:"service"`
	jwt.RegisteredClaims
}

// TokenService handles token generation and validation for RAG server
type TokenService struct {
	publicToken    string
	refreshSecret  string
	accessSecret   string
	accessExpiry   time.Duration
	refreshExpiry  time.Duration
}

// TokenConfig holds configuration for token service
type TokenConfig struct {
	PublicToken     string
	RefreshSecret   string
	AccessSecret    string
	AccessExpiry    time.Duration
	RefreshExpiry   time.Duration
}

// NewTokenService creates a new TokenService instance
func NewTokenService(config TokenConfig) *TokenService {
	return &TokenService{
		publicToken:    config.PublicToken,
		refreshSecret:  config.RefreshSecret,
		accessSecret:   config.AccessSecret,
		accessExpiry:   config.AccessExpiry,
		refreshExpiry:  config.RefreshExpiry,
	}
}

// ValidatePublicToken validates the public token
func (s *TokenService) ValidatePublicToken(publicToken string) bool {
	return publicToken == s.publicToken
}

// GenerateTokenPair generates a new token pair for RAG services
func (s *TokenService) GenerateTokenPair(userID, email string, roles []string) (*TokenPair, error) {
	// Generate refresh token (JWT)
	refreshClaims := jwt.RegisteredClaims{
		Subject:   userID,
		Audience:  []string{"xcontrol-rag-refresh"},
		ExpiresAt: jwt.NewNumericDate(time.Now().Add(s.refreshExpiry)),
		NotBefore: jwt.NewNumericDate(time.Now()),
		Issuer:    "xcontrol-rag",
	}

	refreshToken := jwt.NewWithClaims(jwt.SigningMethodHS256, refreshClaims)
	refreshTokenString, err := refreshToken.SignedString([]byte(s.refreshSecret))
	if err != nil {
		return nil, fmt.Errorf("failed to sign refresh token: %w", err)
	}

	// Generate access token (JWT)
	claims := Claims{
		UserID:  userID,
		Email:   email,
		Roles:   roles,
		Service: "rag-server",
		RegisteredClaims: jwt.RegisteredClaims{
			Subject:   userID,
			Audience:  []string{"xcontrol-rag-access"},
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(s.accessExpiry)),
			NotBefore: jwt.NewNumericDate(time.Now()),
			Issuer:    "xcontrol-rag",
		},
	}

	accessToken := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	accessTokenString, err := accessToken.SignedString([]byte(s.accessSecret))
	if err != nil {
		return nil, fmt.Errorf("failed to sign access token: %w", err)
	}

	return &TokenPair{
		PublicToken:  s.publicToken,
		AccessToken:  accessTokenString,
		RefreshToken: refreshTokenString,
		TokenType:    "Bearer",
		ExpiresIn:    int64(s.accessExpiry.Seconds()),
	}, nil
}

// ValidateAccessToken validates and parses an access token
func (s *TokenService) ValidateAccessToken(accessToken string) (*Claims, error) {
	token, err := jwt.ParseWithClaims(accessToken, &Claims{}, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return []byte(s.accessSecret), nil
	})
	if err != nil {
		return nil, fmt.Errorf("failed to parse access token: %w", err)
	}

	claims, ok := token.Claims.(*Claims)
	if !ok || !token.Valid {
		return nil, fmt.Errorf("invalid access token")
	}

	return claims, nil
}

// RefreshAccessToken generates a new access token using refresh token
func (s *TokenService) RefreshAccessToken(refreshToken string) (string, error) {
	token, err := jwt.ParseWithClaims(refreshToken, &jwt.RegisteredClaims{}, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return []byte(s.refreshSecret), nil
	})
	if err != nil {
		return "", fmt.Errorf("failed to parse refresh token: %w", err)
	}

	claims, ok := token.Claims.(*jwt.RegisteredClaims)
	if !ok || !token.Valid {
		return "", fmt.Errorf("invalid refresh token")
	}

	// Verify issuer and audience
	if claims.Issuer != "xcontrol-rag" {
		return "", fmt.Errorf("invalid token issuer")
	}

	if !contains(claims.Audience, "xcontrol-rag-refresh") {
		return "", fmt.Errorf("invalid token audience")
	}

	// Generate new access token
	newClaims := Claims{
		UserID:  claims.Subject,
		Email:   "",
		Roles:   []string{"user"},
		Service: "rag-server",
		RegisteredClaims: jwt.RegisteredClaims{
			Subject:   claims.Subject,
			Audience:  []string{"xcontrol-rag-access"},
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(s.accessExpiry)),
			NotBefore: jwt.NewNumericDate(time.Now()),
			Issuer:    "xcontrol-rag",
		},
	}

	accessToken := jwt.NewWithClaims(jwt.SigningMethodHS256, newClaims)
	accessTokenString, err := accessToken.SignedString([]byte(s.accessSecret))
	if err != nil {
		return "", fmt.Errorf("failed to sign access token: %w", err)
	}

	return accessTokenString, nil
}

// GetAccessTokenExpiry returns the access token expiry duration
func (s *TokenService) GetAccessTokenExpiry() time.Duration {
	return s.accessExpiry
}

// Helper function to check if a slice contains a string
func contains(slice []string, str string) bool {
	for _, s := range slice {
		if s == str {
			return true
		}
	}
	return false
}

File: dashboard-fresh/lib/auth/token_service.ts
// Token service for Public + Refresh + JWT access_token dual-layer authentication
// Using Deno crypto API

export interface TokenPair {
  public_token: string;
  access_token: string;
  refresh_token: string;
  token_type: string;
  expires_in: number;
}

export interface TokenClaims {
  user_id: string;
  email: string;
  roles: string[];
  service?: string;
  exp: number;
  iat: number;
  iss: string;
  aud: string | string[];
}

/**
 * TokenService handles token generation and validation in the frontend
 */
export class TokenService {
  private accessToken: string | null = null;
  private refreshToken: string | null = null;
  private publicToken: string | null = null;
  private accessSecret: string;
  private refreshSecret: string;

  constructor(config: {
    publicToken: string;
    refreshSecret: string;
    accessSecret: string;
  }) {
    this.publicToken = config.publicToken;
    this.refreshSecret = config.refreshSecret;
    this.accessSecret = config.accessSecret;
  }

  /**
   * Set the public token from configuration
   */
  setPublicToken(token: string): void {
    this.publicToken = token;
  }

  /**
   * Set tokens after successful authentication
   */
  setTokens(tokenPair: TokenPair): void {
    this.accessToken = tokenPair.access_token;
    this.refreshToken = tokenPair.refresh_token;
    this.publicToken = tokenPair.public_token;
    this.storeTokens();
  }

  /**
   * Get the current access token
   */
  getAccessToken(): string | null {
    return this.accessToken;
  }

  /**
   * Get the current refresh token
   */
  getRefreshToken(): string | null {
    return this.refreshToken;
  }

  /**
   * Get the public token
   */
  getPublicToken(): string | null {
    return this.publicToken;
  }

  /**
   * Check if access token is expired
   */
  isTokenExpired(token?: string): boolean {
    const tokenToCheck = token || this.accessToken;
    if (!tokenToCheck) return true;

    try {
      const claims = this.decodeToken(tokenToCheck);
      const currentTime = Math.floor(Date.now() / 1000);
      return claims.exp < currentTime;
    } catch {
      return true;
    }
  }

  /**
   * Decode JWT token without verification (for inspection only)
   */
  decodeToken(token: string): TokenClaims {
    const parts = token.split('.');
    if (parts.length !== 3) {
      throw new Error('Invalid token format');
    }

    const payload = atob(parts[1].replace(/-/g, '+').replace(/_/g, '/'));
    return JSON.parse(payload);
  }

  /**
   * Store tokens in localStorage
   */
  private storeTokens(): void {
    if (typeof window === 'undefined') return;

    if (this.accessToken) {
      localStorage.setItem('access_token', this.accessToken);
    }
    if (this.refreshToken) {
      localStorage.setItem('refresh_token', this.refreshToken);
    }
    if (this.publicToken) {
      localStorage.setItem('public_token', this.publicToken);
    }
  }

  /**
   * Load tokens from localStorage
   */
  loadTokens(): void {
    if (typeof window === 'undefined') return;

    this.accessToken = localStorage.getItem('access_token');
    this.refreshToken = localStorage.getItem('refresh_token');
    this.publicToken = localStorage.getItem('public_token');
  }

  /**
   * Clear all tokens
   */
  clearTokens(): void {
    this.accessToken = null;
    this.refreshToken = null;
    this.publicToken = null;

    if (typeof window !== 'undefined') {
      localStorage.removeItem('access_token');
      localStorage.removeItem('refresh_token');
      localStorage.removeItem('public_token');
    }
  }

  /**
   * Refresh access token using refresh token
   */
  async refreshAccessToken(): Promise<string | null> {
    if (!this.refreshToken) {
      throw new Error('No refresh token available');
    }

    try {
      const response = await fetch('/api/auth/refresh', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          refresh_token: this.refreshToken,
        }),
      });

      if (!response.ok) {
        throw new Error('Failed to refresh token');
      }

      const data = await response.json();
      this.accessToken = data.access_token;
      this.storeTokens();

      return this.accessToken;
    } catch (error) {
      console.error('Token refresh failed:', error);
      this.clearTokens();
      return null;
    }
  }

  /**
   * Create authorization header
   */
  createAuthHeader(token?: string): Record<string, string> {
    const tokenToUse = token || this.accessToken;
    if (!tokenToUse) {
      return {};
    }

    return {
      'Authorization': `Bearer ${tokenToUse}`,
    };
  }

  /**
   * Automatically refresh token if expired
   */
  async ensureValidToken(): Promise<boolean> {
    if (!this.accessToken) {
      return false;
    }

    if (this.isTokenExpired()) {
      const newToken = await this.refreshAccessToken();
      return newToken !== null;
    }

    return true;
  }
}

// Default instance
export const tokenService = new TokenService({
  publicToken: 'xcontrol-public-token-2024',
  refreshSecret: 'xcontrol-refresh-secret-2024',
  accessSecret: 'xcontrol-access-secret-2024',
});

/**
 * Fetch wrapper with automatic token management
 */
export async function authFetch(
  url: string,
  options: RequestInit = {},
  autoRefresh: boolean = true
): Promise<Response> {
  // Ensure token is valid
  if (autoRefresh) {
    await tokenService.ensureValidToken();
  }

  // Add auth header
  const headers = new Headers(options.headers);
  const authHeader = tokenService.createAuthHeader();
  Object.entries(authHeader).forEach(([key, value]) => {
    headers.set(key, value);
  });

  const response = await fetch(url, {
    ...options,
    headers,
  });

  // Handle token expiration
  if (response.status === 401 && autoRefresh) {
    const newToken = await tokenService.refreshAccessToken();
    if (newToken) {
      const newHeaders = new Headers(options.headers);
      const newAuthHeader = tokenService.createAuthHeader(newToken);
      Object.entries(newAuthHeader).forEach(([key, value]) => {
        newHeaders.set(key, value);
      });

      return fetch(url, {
        ...options,
        headers: newHeaders,
      });
    }
  }

  return response;
}
